### PR TITLE
Pivotal ID # 168982339: Project Pattern Sequence Generation

### PR DESCRIPTION
**Pull request by jhoanmanuelms**
- Change the projects endpoint to make the accNo template mandatory and create the corresponding sequence
- Create integration tests for project accNo template
- Move the empty accNo integration tests to submission test class

TODO In Next PR
- Simplify the sequence object by removing the suffix which is unused
- Include project sequences in the test database setup

### CODE REVIEW

File: submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/integration/PersistenceContextImpl.kt [hunk 1]
```diff
@@ -28,6 +29,10 @@ open class PersistenceContextImpl(
     private val subDbMapper: SubmissionDbMapper,
     private val subMapper: SubmissionMapper
 ) : PersistenceContext {
+    override fun createAccNoPatternSequence(pattern: AccPattern) {
+        sequenceRepository.save(Sequence(pattern.prefix))
```
Comment (Juan-EBI) [inline submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/integration/PersistenceContextImpl.kt]: I believe we need to use both prefix and suffix here

File: submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/test/project/submit/ProjectSubmitTest.kt [hunk 2]
```diff
@@ -67,6 +70,7 @@ internal class ProjectSubmitTest(private val tempFolder: TemporaryFolder) : Base
             assertThat(submittedProject.accessTags.first()).isEqualTo("AProject")
 
             assertThat(tagsDataRepository.existsByName("AProject")).isTrue()
+            assertThat(sequenceRepository.existsByPrefixAndSuffix("S-APR", "")).isTrue()
```
Comment (Juan-EBI) [inline submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/test/project/submit/ProjectSubmitTest.kt]: dont we need to validate both?

File: submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/test/submission/submit/SubmissionApiTest.kt [hunk 3]
```diff
@@ -73,6 +81,48 @@ internal class SubmissionApiTest(private val tempFolder: TemporaryFolder) : Base
...
+            )
+        }
+
+        @Test
+        fun `empty accNo with parent`() {
+            val project = tsv {
+                line("Submission", "A-Project")
+                line("AccNoTemplate", "!{S-APR,}")
+                line()
+
```
Comment (Juan-EBI) [inline submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/test/submission/submit/SubmissionApiTest.kt]: any particular reason for this line break

File: submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/submitter/ProjectSubmitter.kt [hunk 4]
```diff
@@ -1,18 +1,25 @@
...
+    private val accNoPatternUtil: AccNoPatternUtil,
     private val validators: List<IProjectValidator>,
     private val processors: List<IProjectProcessor>
 ) {
     fun submit(project: ExtendedSubmission, context: PersistenceContext): Submission {
         validators.forEach { validator -> validator.validate(project, context) }
         processors.forEach { processor -> processor.process(project, context) }
+
+        val sequencePrefix = accNoPatternUtil.getPattern(project.accNoTemplate!!).prefix
+        context.createAccNoPatternSequence(AccPattern(sequencePrefix))
```
Comment (Juan-EBI) [inline submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/submitter/ProjectSubmitter.kt]: should not we use both?

File: submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/integration/PersistenceContextImpl.kt (hunk 1 above)
Comment (jhoanmanuelms) [reply submission/persistence/src/main/kotlin/ac/uk/ebi/biostd/persistence/integration/PersistenceContextImpl.kt]: This won't make sense after I remove the suffix as discussed before.

File: submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/submitter/ProjectSubmitter.kt (hunk 4 above)
Comment (jhoanmanuelms) [reply submission/submitter/src/main/kotlin/ac/uk/ebi/biostd/submission/submitter/ProjectSubmitter.kt]: Same

File: submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/test/submission/submit/SubmissionApiTest.kt (hunk 3 above)
Comment (Juan-EBI) [reply submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/test/submission/submit/SubmissionApiTest.kt]: I mean the single `line()` already represents a line break so not need to add another line break. Do I am missing something?

Comment (jhoanmanuelms) [reply submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/test/submission/submit/SubmissionApiTest.kt]: That's just to make the pagetab more readable. We're doing the same in all the other cases and I think it looks good and makes clear at a single glance, where the line breaks are:

https://github.com/EBIBioStudies/biostudies-backend-services/blob/master/submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/factory/TsvSubmissionFactory.kt#L13

Comment (Juan-EBI) [reply submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/test/submission/submit/SubmissionApiTest.kt]: it looks for me as a double line but is a matter of test. Completely subjective.

Comment (jhoanmanuelms) [reply submission/submission-webapp/src/itest/kotlin/ac/uk/ebi/biostd/itest/test/submission/submit/SubmissionApiTest.kt]: Ok, maybe we should change these and all the others in a separated PR, just to be consistent.

**Pull request merged**